command_cursor.c

```diff
@@ -115,13 +115,19 @@ int php_mongo_enforce_batch_size_on_command(zval *command, int size TSRMLS_DC)
 		add_assoc_zval(command, "cursor", tmp_cursor);
 	}
 
-	if (Z_TYPE_PP(zcursor) != IS_ARRAY) {
-		php_mongo_cursor_throw(mongo_ce_CursorException, NULL, 32 TSRMLS_CC, "The cursor command's 'cursor' element is not an array");
+	if (Z_TYPE_PP(zcursor) != IS_ARRAY && Z_TYPE_PP(zcursor) != IS_OBJECT) {
+		php_mongo_cursor_throw(mongo_ce_CursorException, NULL, 32 TSRMLS_CC, "The cursor command's 'cursor' element is not an array or object");
 		return 0;
 	}
 
 	if (zend_hash_find(HASH_PP(zcursor), "batchSize", 10, (void**) &zbatchsize) == FAILURE) {
-		add_assoc_long(*zcursor, "batchSize", size);
+		zval *tmp_batchsize;
+
+		MAKE_STD_ZVAL(tmp_batchsize);
+		ZVAL_LONG(tmp_batchsize, size);
+
+		/* Use hash functions here, since "cursor" option may be an object */
+		zend_hash_add(HASH_PP(zcursor), "batchSize", 10, &tmp_batchsize, sizeof(zval*), NULL);
```

Comment:
@derickr: This currently segfaults as `*zcursor` ends up pointing to the same memory as `tmp_batchsize` when I get around to testing `listCollections()` below, which predictably fails when we `zend_hash_add()` on `HASH_PP(zcursor)`. Even after stepping through with GDB, I'm quite baffled as to what's causing the problem. Hopefully we can take a look at this next week.